services/simcore-agent/src/simcore_service_simcore_agent/_app.py

```diff
@@ -0,0 +1,94 @@
+import asyncio
+import logging
+import signal
+import traceback
+from contextlib import suppress
+from typing import Any, Callable, Coroutine, Final, Optional
+
+from pydantic import PositiveFloat, PositiveInt
+
+from ._meta import APP_FINISHED_BANNER_MSG, APP_STARTED_BANNER_MSG
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_TASK_WAIT_ON_ERROR: Final[PositiveInt] = 10
+
+
+class Application:
+    HANDLED_EXIT_SIGNALS: set[signal.Signals] = {signal.SIGTERM, signal.SIGINT}
+
+    def __init__(self, loop: Optional[asyncio.AbstractEventLoop] = None):
+        self._loop = asyncio.get_event_loop() if loop is None else loop
+        for sig_name in self.HANDLED_EXIT_SIGNALS:
+            self._loop.add_signal_handler(
+                sig_name,
+                lambda: self._loop.create_task(self._exit_gracefully(self._loop)),
+            )
+
+        self._keep_running: bool = True
+
+        self._tasks: set[asyncio.Task] = set()
+        self._coroutines: set[Coroutine] = set()
+        self._registered_coroutines: set[str] = set()
+
+    async def _exit_gracefully(self, loop: asyncio.AbstractEventLoop):
+        for task in self._tasks:
+            task.cancel()
```

Comment:
suggestion, this could be done using asyncio.gather instead. it would probably make stopping slightly faster.